### PR TITLE
Fix: email change improvements

### PR DESCRIPTION
**Pull request by kangmingtay**
## What kind of change does this PR introduce?
Addresses #60  and #78 

## What is the current behavior?
Email change verification is broken.

## What is the new behavior?
Allows a user to change their email. Requires the user's jwt access token to call `PUT /user`.

For security reasons, we send out 2 emails when a user makes an update email request:
1. Confirm email change email is sent to user's current email
2. Confirm email change email is sent to user's new email

The user needs to log into both email accounts and verify by clicking both links.

## Design
`email_change_token_current`: Keeps track of token sent to current email
`email_change_token_new`: Keeps track of token sent to new email
`email_change_confirm_status`: Keeps track of the number of verified email (0, 1, 2)

## Additional context
Adds `/verify?type=email_change` to handle email change verification logic.

### CODE REVIEW

File: mailer/template.go [hunk 1]
```diff
@@ -112,32 +112,57 @@ func (m *TemplateMailer) ConfirmationMail(user *models.User, referrerURL string)
...
+				"Data":            user.UserMetaData,
+			}
+			errors <- m.Mailer.Mail(
+				e,
+				string(withDefault(m.Config.Mailer.Subjects.EmailChange, "Confirm Email Change")),
+				m.Config.Mailer.Templates.EmailChange,
+				defaultEmailChangeMail,
+				data,
+			)
+		}(email, token)
```
Comment (kangmingtay) [inline mailer/template.go:155]: Send emails concurrently via goroutines since they don't have to be sequential

File: mailer/template.go [hunk 2]
```diff
@@ -112,32 +112,57 @@ func (m *TemplateMailer) ConfirmationMail(user *models.User, referrerURL string)
...
-		data,
-	)
+	for i := 0; i < len(tokens); i++ {
+		e := <-errors
+		if e != nil {
+			return e
+		}
+	}
+
+	return nil
```
Comment (kangmingtay) [inline mailer/template.go:165]: check if any of the goroutines experienced an error sending the email

File: models/user.go [hunk 3]
```diff
@@ -341,6 +352,11 @@ func FindUserByRecoveryToken(tx *storage.Connection, token string) (*User, error
 	return findUser(tx, "recovery_token = ?", token)
 }
 
+// FindUserByRecoveryToken finds a user with the matching recovery token.
```
Comment (thebengeu) [inline models/user.go:355]: Update comment?

File: migrations/20210730183235_add_email_change_confirmed.up.sql [hunk 4]
```diff
@@ -0,0 +1,15 @@
+-- adds email_change_confirmed
+
+ALTER TABLE auth.users
+ADD COLUMN IF NOT EXISTS email_change_token_current varchar(255) null DEFAULT '', 
+ADD COLUMN IF NOT EXISTS email_change_confirm_status smallint DEFAULT 0 CHECK (email_change_confirm_status >= 0 AND email_change_confirm_status <= 2);
+
+DO $$
+BEGIN
+  IF NOT EXISTS(SELECT *
```
Comment (thebengeu) [inline migrations/20210730183235_add_email_change_confirmed.up.sql:9]: Should this be `IF EXISTS`?

File: migrations/20210730183235_add_email_change_confirmed.up.sql [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    FROM information_schema.columns
+    WHERE table_schema = 'auth' and table_name='users' and column_name='email_change_token')
+  THEN
+      ALTER TABLE "auth"."users" RENAME COLUMN "email_change_token" TO "email_change_token_new";
+  END IF;
+END $$;
```
Comment (thebengeu) [inline migrations/20210730183235_add_email_change_confirmed.up.sql:15]: Nit: add newline at end of file?

**Review by thebengeu — APPROVED** [review]

File: migrations/20210730183235_add_email_change_confirmed.up.sql (hunk 4 above)
Comment (kangmingtay) [reply migrations/20210730183235_add_email_change_confirmed.up.sql:9]: ah yeah you're right, i forgot to remove my docker volume after making the change manually so i thought the migration worked 😅

**Pull request merged**